### PR TITLE
Reimplementation of NestedExtensionArray

### LINKED ISSUE #15: Do not depend on internal pandas API
Currently, our `NestedExtendedArray` inherits from `ArrowExtendedArray`. Some overrides require usage of pandas internal (protected) classes. This design has multiple issues:

1. We are required to pin pandas version and rely to unstable interfaces: `ArrowExtendedArray` itself is marked unstable.
2. We can make code more clean and performant without `ArrowExtendedArray`
3. We have less control over array behavior. Hopefully we can solve the issue of an assignment of a dataframe to an item this way

### LINKED ISSUE #24: Proper support of missed values
**Bug report**

Current implementation of `NestedExtensionArray` and `.nest` accessor do not handle missed values properly. We should fix it with a good test coverage.

**Before submitting**
Please check the following:

- [x] I have described the situation in which the bug arose, including what code was executed, information about my environment, and any applicable data others will need to reproduce the problem.
- [ ] I have included available evidence of the unexpected behavior (including error messages, screenshots, and/or plots) as well as a descriprion of what I expected instead.
- [ ] If I have a solution in mind, I have provided an explanation and/or pseudocode and/or task list.

### PR DESCRIPTION
**Pull request by hombit**
This reimplements `NestedExtensionArray` inheriting it directly from `ExtensionArray` instead of `ArrowExtensionArray`. It removes dependency on unstable `ArrowExtensionArray`, gives us more control over the behavior, and allows some performance optimization.

If fixes #15, fixes #24 

## Change Description

- [ ] My PR includes a link to the issue that I am addressing

## Code Quality
- [ ] I have read the Contribution Guide
- [ ] My code follows the code style of this project
- [ ] My code builds (or compiles) cleanly without any errors or warnings
- [ ] My code contains relevant comments and necessary documentation

### Bug Fix Checklist
- [ ] My fix includes a new test that breaks as a result of the bug (if possible)
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### New Feature Checklist
- [ ] I have added or updated the docstrings associated with my feature using the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html)
- [ ] I have updated the tutorial to highlight my new feature (if appropriate)
- [ ] I have added unit/End-to-End (E2E) test cases to cover my new feature
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### Documentation Change Checklist
- [ ] Any updated docstrings use the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html)

### Build/CI Change Checklist
- [ ] If required or optional dependencies have changed (including version numbers), I have updated the README to reflect this
- [ ] If this is a new CI setup, I have added the associated badge to the README

### Other Change Checklist
- [ ] Any new or updated docstrings use the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to highlight my new feature (if appropriate)
- [ ] I have added unit/End-to-End (E2E) test cases to cover any changes
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### CODE REVIEW

**Review by dougbrn — APPROVED** [review]
Looks good, just a couple nits/questions that you can feel free to address or ignore!

File: src/nested_pandas/series/ext_array.py [hunk 1]
```diff
@@ -40,52 +107,463 @@ class NestedExtensionArray(ArrowExtensionArray):
...
-        self._dtype = NestedDtype.from_pandas_arrow_dtype(self._dtype)
+        Parameters
+        ----------
+        scalars : Sequence
+            The sequence of scalars: disctionaries, DataFrames, None, pd.NA, pa.Array or anything convertible
+            to PyArrow scalars.
+        dtype : dtype or None
+            dtype of the resulting array
+        copy : bool
+            Ignored, because PyArrow arrays are immutable.
```
[the remark this answers is not in the document]
Comment (hombit) [reply src/nested_pandas/series/ext_array.py:124]: It is in the signature of the base method, so we must have it. Actually, it is questionable if we need support it here, I'd appreciate this discussion

File: src/nested_pandas/series/ext_array.py [hunk 2]
```diff
@@ -40,52 +107,463 @@ class NestedExtensionArray(ArrowExtensionArray):
...
+        if (indices_array < 0).any():
+            # Don't modify in-place
+            indices_array = np.copy(indices_array)
+            indices_array[indices_array < 0] += len(self._pa_array)
+        return type(self)(self._pa_array.take(indices_array))
+
+    def copy(self) -> Self:  # type: ignore[name-defined] # noqa: F821
+        """Return a copy of the extension array.
+
+        This implementation returns a shallow copy of the extension array,
```
[the remark this answers is not in the document]
Comment (hombit) [reply src/nested_pandas/series/ext_array.py:389]: It is a good question. `ArrowExtensionArray` doesn't copy it, and I *think* we should do the same. I'd be happy to discuss more on it

File: src/nested_pandas/series/ext_array.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dougbrn) [reply src/nested_pandas/series/ext_array.py:124]: Right, I'm mainly worried about a user trying to set a flag that does nothing. But at the same time, you've clearly explained here in the docstring that this doesn't do anything, so perhaps that would be on them and fine.

File: src/nested_pandas/series/ext_array.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dougbrn) [reply src/nested_pandas/series/ext_array.py:389]: Yeah, regardless I think it's fine to move forward without it for now

File: src/nested_pandas/series/ext_array.py [hunk 3]
```diff
@@ -1,27 +1,94 @@
...
+    if isinstance(dtype, pd.ArrowDtype):
+        return dtype.pyarrow_dtype
+    if isinstance(dtype, pa.DataType):
+        return dtype
+    return None
+
+
+def replace_with_mask(array: pa.ChunkedArray, mask: pa.BooleanArray, value: pa.Array) -> pa.ChunkedArray:
+    """Replace the elements of the array with the value where the mask is True"""
+    # TODO: performance optimization
```
Comment (wilsonbb) [inline src/nested_pandas/series/ext_array.py:70]: Can we have an issue to investigate this?

File: src/nested_pandas/series/ext_array.py [hunk 4]
```diff
@@ -40,52 +107,463 @@ class NestedExtensionArray(ArrowExtensionArray):
...
+
+        scalar_or_array = self._pa_array[item]
+        if isinstance(scalar_or_array, pa.StructScalar):
+            return self._convert_struct_scalar_to_df(scalar_or_array, copy=False)
+        # Logically, it must be a pa.ChunkedArray if it is not a scalar
+        pa_array = cast(pa.ChunkedArray, scalar_or_array)
+        return type(self)(pa_array, validate=False)
+
+    def __setitem__(self, key, value) -> None:
+        # TODO: optimize for many chunks:
```
Comment (wilsonbb) [inline src/nested_pandas/series/ext_array.py:172]: Adding an issue for this as well might be helpful

**Review by wilsonbb — APPROVED** [review]
This looks good to me, though I'm not very knowledgeable about arrow so there might be some aspects that I'm missing

One thing I'm curious about given the size of the change is if we can go ahead re-run  @dougbrn's benchmarking notebook, but I don't think it should strictly block this PR

File: src/nested_pandas/series/ext_array.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hombit) [reply src/nested_pandas/series/ext_array.py:124]: It is also an internal method, there is a user-level wrapping method `from_sequence` which does not have a copy argument

**Comment by dougbrn** [comment]
@wilsonbb @hombit I've run my benchmarking notebook on this branch and it looks like performance is basically identical.

**Pull request merged**